feat: Auction 내부에서 경매 지속시간에 대한 검증을 진행합니다.

HiiWee opened the pull request:
## 📄 Summary

- 기존 AuctionService에서 진행하던 코드를 옮김
- 나노 세컨드 까지 검증하도록 변경
- 겸애 지속시간 검증을 `분 -> 나노 초`로 변경했을때 경매 지속시간이 `60.00000xx분`과 같은 차이로 조건에 부합하지 않아 실패한 테스트를 수정 합니다.
- `PercentagePricePolicy`, `ConstantPricePolicy`에서 경매 지속시간과, 가격 감소 횟수로 계산했을때 경매가격이 0원 이하로 떨어지면 예외가 발생하는 검증을 추가했습니다.
  -  검증 코드는 `지속시간 / 가격하락주기 시간`이 나누어 떨어졌을때를 가정하고 작성했습니다.
  -  검증 추가로 인해 실패하는 테스트 케이스도 수정했습니다.
- 그 외로 Fixture 적용할 수 있는 테스트는 적용하여 코드볼륨을 조오금 줄이고자 했습니다!

## 🙋🏻 More
- 테스트를 수정하면서 느낀거지만, 현재 `AuctionService`에서 `Auction`의 `currentStatus()`를 통해 상태를 꺼내서 검증하고 있기 때문에
  `Auction`객체를 생성할때 이미 시작 중인 상태에서 `changeStock` 메소드를 테스트한다거나 하는 상태의 불일치가 보이는것 같아요! 
  그래서 `Auction` 도메인 내부로  현재 `status`에 따라 해당 메소드들을 실행 시킬 수 있도록 검증하는 부분을 옮기는게 어떨까 하는 생각이 듭니다!

close #169

HiiWee committed:
feat: Auction 내부에서 경매 지속시간에 대한 검증을 진행합니다.

- 기존 AuctionService에서 진행하던 코드를 옮김
- 나노 세컨드 까지 검증하도록 변경
- 조건에 부합하지 않았던 테스트 수정

HiiWee commented on src/main/java/com/wootecam/luckyvickyauction/core/auction/service/AuctionService.java:
```diff
@@ -18,24 +18,17 @@
 import java.util.List;
 import lombok.RequiredArgsConstructor;
 
+// TODO: [update, change 메소드에서 경매 상태 조건을 확인하는 부분을 서비스가 아니라 Auction이 갖게 하도록 변경하기] [writeAt: 2024/08/15/14:18] [writeBy: HiiWee]
 @RequiredArgsConstructor
```
More에 적은 부분에 해당되는 주석입니다!

chhs2131 replied:
좋은 것 같습니다. 흐름상으로도 도메인 측에 물어보는게 더 자연스러울 것 같아요~

HiiWee committed:
Merge branch 'dev' into feat/169-issue

minseok-oh commented on src/main/java/com/wootecam/luckyvickyauction/core/auction/domain/Auction.java:
```diff
@@ -56,8 +59,18 @@ public Auction(
         this.isShowStock = isShowStock;
+    }
 
-        pricePolicy.validate(originPrice);
+    private void validateDurationTime(ZonedDateTime startedAt, ZonedDateTime finishedAt) {
+        Duration diff = Duration.between(startedAt, finishedAt);
+        long diffNanos = diff.toNanos();
+        long tenMinutesInNanos = 10L * 60 * 1_000_000_000; // 10분을 나노초로 변환
+
+        if (!(diffNanos % tenMinutesInNanos == 0 && diffNanos / tenMinutesInNanos <= 6)) {
+            String message = String.format("경매 지속 시간은 10분 단위여야하고, 최대 60분까지만 가능합니다. 현재: %.9f분",
+                    diffNanos / (60.0 * 1_000_000_000));
+            throw new BadRequestException(message, ErrorCode.A008);
+        }
     }
```
세상에... 고생하셨습니다...

minseok-oh commented on src/test/java/com/wootecam/luckyvickyauction/core/auction/domain/AuctionTest.java:
```diff
@@ -59,56 +60,24 @@ public class PercentagePricePolicyTest {
-    public class UpdateAuctionTest {
+    public class updateShowStock_메소드는 {
 
         @Test
-        @DisplayName("동일한 판매자는 현재 경매의 가격 노출 정책을 변경할 수 있다.")
-        public void updateAuctionShowStock() {
+        public void 동일한_판매자는_현재_경매의_가격_노출_정책을_변경할_수_있다() {
             // given
```
의견: "는"이 2번이라 밑의 테스트와 비슷하게 동일한_판매자인_경우 는 어떨까요? 취향차이~ ㅎㅎ

HiiWee replied:
아 이거 후다닥 바꾸다보니,, 뽀록나버렸다 수정할게요,,

minseok-oh commented on src/main/java/com/wootecam/luckyvickyauction/core/auction/domain/Auction.java:
```diff
@@ -43,6 +43,9 @@ public Auction(
             boolean isShowStock
     ) {
+        validateDurationTime(startedAt, finishedAt);
+        pricePolicy.validate(originPrice);
+
```
질문 : 총 시간동안 할인된 금액이 0원 이하인 경우는 이 함수에서 예외를 터뜨리게끔 구현되는 건가요?

HiiWee replied:
아니요! 깜박했네요 추가할게요 고마워요 민석쿤😉

minseok-oh reviewed (dismissed):
코드 잘봤습니다!! 조그마한 질문이 있어서 남기고 갑니당

yudonggeun commented:
Summery -> 조건에 부합하지 않았던 테스트 수정에서 그 조건이 무엇인가요?

yudonggeun reviewed (dismissed):
👍 👍 👍 👍 👍 👍 👍 👍

HiiWee commented:
> Summery -> 조건에 부합하지 않았던 테스트 수정에서 그 조건이 무엇인가요?

날카롭다,,!! 조건 주어 추가했습니다 ㅎ 나중에 보면 이해 안될뻔했네요 ㅠ

chhs2131 reviewed (dismissed):
피자 피자 피자 맛있게 드세요!!!!  

![image](https://github.com/user-attachments/assets/ebb59087-6635-4740-8f46-67d41ae25aa2)

HiiWee committed:
feat: 고정 가격, 퍼센트 가격 감소 정책에서 최대 할인가격을 적용했을때 0원이 되지 않는지 검증한다.

- 60분동안 10분마다 가격을 내릴때 마지막 6번째 가격하락은 이미 경매가
      종료됐으므로 검증 범위에서 제외합니다
- 퍼센트 가격 감소는 퍼센트를 적용하고 항상 소수점 자리를 버립니다.

HiiWee committed:
Merge branch 'dev' into feat/169-issue

HiiWee commented:
`Constant, PercentagePricePolicy`에서 경매 가격이 0원까지 떨어졌을때를 대비한 검증을 추가했습니다!

yudonggeun commented on src/main/java/com/wootecam/luckyvickyauction/core/auction/domain/PercentagePricePolicy.java:
```diff
@@ -15,22 +20,29 @@ public PercentagePricePolicy(double discountRate) {
-            throw new BadRequestException(
-                    String.format("상품 원가는 할인율을 적용한 가격보다 커야 합니다. 상품 원가: %d, 할인율: %f", price, discountRate),
-                    ErrorCode.A010);
+    public void validate(ZonedDateTime startedAt, ZonedDateTime finishedAt, Duration variationDuration, long price) {
+        Duration totalDuration = Duration.between(startedAt, finishedAt);
+        int numberOfVariations = (int) totalDuration.dividedBy(variationDuration);
+
+        long discountedPrice = price;
+        double discountFactor = (100 - discountRate) / 100.0;
+
+        for (int i = 0; i < numberOfVariations - 1; i++) {
+            discountedPrice = (long) Math.floor(discountedPrice * discountFactor);
+            if (discountedPrice <= 0) {
+                String message = String.format("경매 진행 중 가격이 0원 이하가 됩니다. 초기 가격: %d, 할인횟수: %d, 할인율: %d%%", price,
+                        numberOfVariations - 1, (int) discountRate);
+                throw new BadRequestException(message, ErrorCode.A029);
+            }
```
할인 가격이 0이 되지 않아야하는 검증 책임을 PricePolicy에서 처리하는 것 보다. Auction 갱신 메서드에서 확인을 해주는 것이 좋다고 생각해요.

#### 이유
* 가격 정책이 복잡한 로직을 구현해서 복잡도가 올라간다.
* 경매 진행 중간에 최대 할인을 적용하고 싶은 요구 사항이 있을 수 있다.

HiiWee replied:
기존 가격 정책을 보면 크게 검증을 많이 해주는 부분이 없다고 느껴졌었어요 그래서 일단 내부에 구현하긴 했습니당
(추가로 Auction 내부에서 진행하면 policy의 구현체까지 알아내고 알맞은 검증을 해주어야 해야해서 꽤 복잡하다고 느껴졌습니당)

가격 정책에 대한 검증을 Auction내부에서 진행하면 Policy를 분리하지 않고 단순히 Enum값으로 관리해도 괜찮지 않을까요?!

yudonggeun replied:
가격 정책은 가격의 할인하는 조건에 따라서 계산하는 방식이 달라지는 방식이라서 Enum 보다는 Interface를 사용했어요. 일급함수를 사용해서 구현하는 방식이라면 Enum으로도 구현할 수 있다고 생각해서 좋아요.

가격정책은 다음 주기의 가격이 어떻게 되는지를 계산하는 역할만 가지면 된다고 생각을 했어요. 그런 역할을 수행한다는 가정하에 가격 갱신 과정은 다음과 같을 것 같아요.

#### 가격 갱신 과정
1. Auction의 currentPrice, originPrice를 조회한다.
2. PricePolicy를 적용한 다음 가격을 계산한다.
    * 여기에서는 PricePolicy의 구현에 대해서 추상화한다.
3. Auction에서 현재가격을 갱신한다.
    3-1. 만약 계산한 가격이 0(최소 가격 한계선)이하이면 갱신에 실패한다.
    3-2 성공시에는 가격을 갱신한다.

이 방식의 장점은 가격 갱신 과정에서 시간에 대한 정보를 고려하지 않아도 되어서 복잡도가 조금 낮아지는 방향이지 않은가? 하는 생각입니다.

yudonggeun reviewed (commented):
👍

yudonggeun commented:
- #169 가 어느 스토리에 연관된 이슈인지 매핑해주실 수 있을까요???

HiiWee commented:
> * [[FEATURE] 경매 도메인 생성 시 지속시간을 검증하도록 한다. #169](https://github.com/woowa-techcamp-2024/Team7-ELEVEN/issues/169) 가 어느 스토리에 연관된 이슈인지 매핑해주실 수 있을까요???

경매 생성 한다라는 스토리에 어울리는데, 지금 스토리 중에 딱히 매핑하기 알맞은게 없어서 일단 제외해놨습니다. 새로 만드는게 나을까여?

yudonggeun commented:
확실히 스토리가 있으면 문맥 파악하기 좋아서 `판매자는 경매를 등록(생성)할 수 있다.`로 스토리를 만들면 좋을 것 같습니다!!

yudonggeun commented:
- #191 추가해서 매핑했습니다!

HiiWee committed:
refactor: pricePolicy에서 진행하던 경매 진행 중 최소값 검증 로직을 Auction내부로 옮긴다.

- Auction 생성 시 검증하도록 변경했습니다.

HiiWee committed:
Merge branch 'dev' into feat/169-issue

yudonggeun reviewed (approved):
머지 드가자!!
